cmd/immuclient/cli/cli.go

```diff
@@ -0,0 +1,284 @@
+	}
+
+	cli.valueOnly = viper.GetBool("value-only")
+	cli.commands = make(map[string]*command)
+	cli.commandsList = make([]*command, 0)
+	return cli, nil
+}
+
+func options() *client.Options {
+	port := viper.GetInt("immudb-port")
+	address := viper.GetString("immudb-address")
+	authEnabled := viper.GetBool("auth")
+	tokenFileName := viper.GetString("tokenfile")
+	mtls := viper.GetBool("mtls")
+	certificate := viper.GetString("certificate")
+	servername := viper.GetString("servername")
+	pkey := viper.GetString("pkey")
+	clientcas := viper.GetString("clientcas")
+	options := client.DefaultOptions().
+		WithPort(port).
+		WithAddress(address).
+		WithAuth(authEnabled).
+		WithTokenFileName(tokenFileName).
+		WithMTLs(mtls)
+	if mtls {
+		// todo https://golang.org/src/crypto/x509/root_linux.go
+		options.MTLsOptions = client.DefaultMTLsOptions().
+			WithServername(servername).
+			WithCertificate(certificate).
+			WithPkey(pkey).
+			WithClientCAs(clientcas)
+	}
+	return options
+}
+
+func (cli *cli) Register(cmd *command) {
+	cli.commandsList = append(cli.commandsList, cmd)
+	cli.commands[cmd.name] = cmd
+}
+
+func (cli *cli) helpInit() {
+	var namelen, shortlen int
+	name := make([]string, 0)
+	short := make([]string, 0)
+	args := make([]string, 0)
+	for i := range cli.commandsList {
+		if len(cli.commandsList[i].name) > namelen {
+			namelen = len(cli.commandsList[i].name)
+		}
+		if len(cli.commandsList[i].short) > shortlen {
+			shortlen = len(cli.commandsList[i].short)
+		}
+		name = append(name, cli.commandsList[i].name)
+		short = append(short, cli.commandsList[i].short)
+		if len(cli.commandsList[i].args) == 0 {
+			args = append(args, "")
+		} else {
+			args = append(args, strings.Join(cli.commandsList[i].args, ","))
+		}
+	}
+	str := strings.Builder{}
+	for i := range name {
+		str.WriteString(padRight(name[i], " ", namelen+2))
+		str.WriteString(padRight(short[i], " ", shortlen+2))
+		if len(args[i]) > 0 {
+			str.WriteString("args: " + args[i])
+		}
+		str.WriteString("\n")
+	}
+	str.WriteString("\n")
+	cli.helpMessage = str.String()
+}
+
+func (cli *cli) Run() {
+	cli.initCommands()
+	cli.helpInit()
+	l, err := readline.New("immuclient>")
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return
+	}
+	defer l.Close()
+	for {
+		line, err := l.Readline()
+		if err == readline.ErrInterrupt {
+			if len(line) == 0 {
+				break
+			} else {
+				continue
+			}
+		} else if err == io.EOF {
+			continue
+		}
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		line = strings.TrimSuffix(line, "\n")
+		arrCommandStr := strings.Fields(line)
+		if len(arrCommandStr) == 0 {
+			continue
+		}
+		passed := cli.checkCommand(arrCommandStr)
+		if passed {
+			cli.runCommand(arrCommandStr)
+		}
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+		}
+	}
+}
+
+func (cli *cli) checkCommand(arrCommandStr []string) bool {
+	if arrCommandStr[0] == "exit" || arrCommandStr[0] == "quit" {
+		if cli.isLoggedin {
+			logoutmsg, _ := cli.logout(nil)
```

Comment:
@volknanebo Why is it needed to delete the token file (by calling `logout`) and force this way the user to re-authenticate after he exits the interactive mode?

Reply:
I've discussed with Dennis and he thinks to prefer it to be this way. So nevermind my question then.